ramips: add support for Sitecom WLR-4100 v1 002

polex opened the pull request:
Sitecom WLR-4100 v1 002 (marked as X4 N300) is a wireless router

**Specification:**
SoC: MT7620A
RAM: 64 MB DDR2
Flash: MX25L6405D SPI NOR 8 MB
WIFI: 2.4 GHz integrated
Ethernet: 5x 10/100/1000 Mbps QCA8337
USB: 1x 2.0
LEDS: 2x GPIO controlled, 5x switch
Buttons: 1x GPIO controlled
UART: row of 4 unpopulated holes near USB port, starting count from white triangle on PCB:
1. VCC 3.3V
2. GND
3. TX
4. RX

- baud: 115200, parity: none, flow control: none

**Installation**
1. Connect to one of LAN (yellow) ethernet ports,
2. Open router configuration interface,
3. Go to Toolbox > Firmware,
4. Browse for OpenWrt factory image with dlf extension and hit Apply,
5. Wait few minutes, after the Power LED will stop blinking, the router is ready for configuration.

**Known issues**
    Some USB 2.0 devices work at full speed mode 1.1 only

**Additional information**
    OEM firmware shell password is: SitecomSenao
    useful for creating backup of original firmware.
    There is also another revision of this device (v1 001), based on RT3352 SoC

Signed-off-by: Andrea Poletti <polex73@yahoo.it>

mpratt14 commented:
the commit itself needs a title just like the PR title

`ramips: add support for Sitecom WLR-4100 v1 002`

note the spelling of "ramips", I see this mistake from several people
it is short for "Ralink MIPS"

mpratt14 commented:
put the model of the flash chip in the commit message
if you cannot open the case to read the label of the chip, you can get it from the kernel log

mpratt14 commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,19 @@ define Device/sercomm_na930
+define Device/sitecom_wlr-4100v1002
+  SOC := mt7620a
+  BLOCKSIZE := 4k
+  IMAGE_SIZE := 7244k
```
hex to decimal
0x79 --> 121

blocks * blocksize
121 * 64k = 7744k

mpratt14 commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,19 @@ define Device/sercomm_na930
+  BLOCKSIZE := 4k
+  IMAGE_SIZE := 7244k
+  IMAGES += factory.dlf
+  IMAGE/factory.dlf := $$(sysupgrade_bin) | check-size $$$$(IMAGE_SIZE) | \
```
remove `$$$$(IMAGE_SIZE)`, this is the default of `check-size`

mpratt14 commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,19 @@ define Device/sercomm_na930
+	senao-header -r 0x0222 -p 0x104A -t 2
+  DEVICE_VENDOR := Sitecom
+  DEVICE_MODEL := WLR-4100v1002
+  DEVICE_PACKAGES := kmod-usb-ohci
```
include `kmod-usb2` for usb 2.0

mpratt14 commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -133,6 +133,7 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch1" \
 			"1:lan" "2:lan" "3:lan" "4:lan" "0:wan" "6@eth0"
 		;;
+	sitecom,wlr-4100v1002|\
```
alphabetical order

this means changing engenius,esr600 to have the `|\` and your board have the `)`

mcprat reviewed (commented):
👍🏼

mpratt14 commented:
we need to do something about the board name since it has 2 version numbers

something like `v1-002` or `v1_002`

I don't know which one is better

mpratt14 commented:
@polex if you need help updating PR just ask 👍🏼

mpratt14 commented:
@polex also I noticed a lot of the talk on your original forum post is about configuring the RGMII interface

I added some info to the wiki and made a test image for TFTP boot that prints all the registers of the switch
hope it is helpful

https://openwrt.org/docs/techref/hardware/soc/soc.mediatek#rgmii_configuration
https://openwrt.org/docs/techref/hardware/switch/atheros

https://gist.github.com/mpratt14/d4bd90138c310ec745a67d354058af1c

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "sitecom,wlr-4100v1002", "ralink,mt7620a-soc";
```
Do you really think this is a good spelling?

polex replied:
I have no idea. What do you suggest?

adschm replied:
Well, replacing the spaces with hyphens, like in all of the other examples?
sitecom,wlr-4100-v1-002
But I still doubt we actually need the 002 here ...

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+
+/ {
+	compatible = "sitecom,wlr-4100v1002", "ralink,mt7620a-soc";
+	model = "Sitecom sw66 WLR-4100 v1 002";
```
What's with the stray sw66?

polex replied:
Sorry, It is a refusal of several "*.dts" file versions

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+	
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
```
add an empty line ...

polex replied:
OK

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+	};
+};
+
+
```
remove second empty line.

adschm replied:
Resolved, but not changed.....

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+					0x04 0x06200000  // PORT0 PAD MODE CTRL
+					0x08 0x01000000  //  PORT5 PAD MODE CTRL  RX delay EN all ports 0, 5, 6 
+					0x7C 0x0000004e  // 	PORT0_STATUS 
+					0x94 0x00000000   // PORT6_STATUS
```
broken indent/formatting.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+		status = "okay";
+		mediatek,mdio-mode = <1>;
+		phy0: ethernet-phy@0 {
+				reg = <0>;
```
broken indent.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+		mediatek,mdio-mode = <1>;
+		phy0: ethernet-phy@0 {
+				reg = <0>;
+				phy-mode = "rgmii";
```
add empty line after this.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+};
+
+&wmac {
+	status = "okay";
```
add empty line.

polex replied:
OK

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+
+&state_default {
+	gpio {
+			groups = "uartf", "i2c", "rgmii2", "wled", "spi refclk", "ephy",  "nd_sd";
```
you have more groups that gpios used, so some must be wrong ...

polex replied:
Actually I listed all pins which were not used for specific function.
I tried to leave WiFi LED (GPIO 72) driven by HW but it did not work. So I left it driven by SW like all other devices.

adschm replied:
Please only use those that you actually need ...

adschm commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,19 @@ define Device/sercomm_na930
+  IMAGE/factory.dlf := $$(sysupgrade_bin) | check-size $$$$(IMAGE_SIZE) | \
+	senao-header -r 0x0222 -p 0x104A -t 2
+  DEVICE_VENDOR := Sitecom
+  DEVICE_MODEL := WLR-4100v1002
```
BTW: do we really need the 002? Is it different from the 001?

polex replied:
v1 001 is based on different SoC (RT3352)

adschm commented on target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:
```diff
@@ -131,6 +131,7 @@ kimax,u35wf)
 	;;
 kingston,mlw221|\
 kingston,mlwg2|\
+sitecom,wlr-4100v1002|\
```
Wifi trigger can be moved to the DTS file.

polex replied:
Sorry. I have no idea on how to embed it *.dts

mpratt14 replied:
@polex wifi LED trigger in DTS like this...

```
		wlan2g {
			label = "blue:wlan2g";
			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
			linux,default-trigger = "phy1tpt";
		};
```

change to "phy0tpt" for phy0
if it still does not work then do "phy0radio"

adschm replied:
In general, it's a good idea to look at other device files before/while compiling a PR ...

adschm commented:
Comments above plus the formal issue already reported regarding the commit title ...

polex commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+		led_power: power {	
+			label = "amber:power";
+			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
+		};
```
OK

adschm replied:
... and move it after the gpios line ...

polex commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,184 @@
+&state_default {
+	gpio {
+			groups = "uartf", "i2c", "rgmii2", "wled", "spi refclk", "ephy",  "nd_sd";
+			function = "gpio";
```
You are right

mpratt14 commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,181 @@
+	leds {
+		compatible = "gpio-leds";
+		
+		led_power:power {	
```
this label now broken since last change, needs the space `led_power: power {`

mpratt14 commented:
I think board "compatible" should be `sitecom,wlr-4100_v1-002`

unless you think of something better @adschm 

in makefiles the variant is separated by dash `-` in device definition but I'm thinking it would be more consistent the other way
to have VENDOR_MODEL_VARIANT
instead of VENDOR_MODEL-VARIANT

mpratt14 commented:
@polex if you need help combining your commits into 1 commit, let me know

adschm commented:
> I think board "compatible" should be `sitecom,wlr-4100_v1-002`
> 
> unless you think of something better @adschm
> 
> in makefiles the variant is separated by dash `-` in device definition but I'm thinking it would be more consistent the other way
> to have VENDOR_MODEL_VARIANT
> instead of VENDOR_MODEL-VARIANT

Nope, it's vendor,model-variant, so sitecom,wlr-4100-v1-002

adschm commented:
Please squash all your changes into a single, proper commit.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,181 @@
+
+&gpio1 {
+	status = "okay";
+};
```
Add empty line.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,181 @@
+	};
+
+	mdio-bus {
+		status = "okay";
```
Add empty line.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100v1002.dts:
```diff
@@ -0,0 +1,181 @@
+	mtd-mac-address = <&factory 0x4>;
+
+	port@5 {
+		status = "okay";
```
Add empty line after status.

adschm commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,20 @@ define Device/sercomm_na930
+	senao-header -r 0x0222 -p 0x104A -t 2
+  DEVICE_VENDOR := Sitecom
+  DEVICE_MODEL := WLR-4100
+  DEVICE_VARIANT := v1-002
```
Remove hyphen for variant.

adschm replied:
No, "v1 002" with space instead of hyphen

mpratt14 commented:
does the label really have "1 002" as the version or just 2

polex commented:
The label reports WLR-4100 v1 002|X4 N300
![wlr41000stick](https://user-images.githubusercontent.com/1233479/126452942-55d1d0d6-b469-4bea-a5f4-6b80131d5af0.jpg)
I am not aware of any other WLR-4100 variants except V1 001 which is based on different SoC (major release):
https://forum.openwrt.org/t/openwrt-support-for-sitecom-wlr-4100-v1-001/73193
So we may distinguish them by just the last version digit (v1 vs. v2).

polex committed:
ramips: add support for Sitecom wlr-4100 v1 002

Sitecom WLR-4100 v1 002 (marked as X4 N300) is a wireless router
Specification:
SoC: MT7620A
RAM: 64 MB DDR2
Flash: MX25L6405D SPI NOR 8 MB
WIFI: 2.4 GHz integrated
Ethernet: 5x 10/100/1000 Mbps QCA8337
USB: 1x 2.0
LEDS: 2x GPIO controlled, 5x switch
Buttons: 1x GPIO controlled
UART: row of 4 unpopulated holes near USB port, starting count from white triangle on PCB:

    VCC 3.3V
    GND
    TX
    RX

    baud: 115200, parity: none, flow control: none

Installation

    Connect to one of LAN (yellow) ethernet ports,
    Open router configuration interface,
    Go to Toolbox > Firmware,
    Browse for OpenWrt factory image with dlf extension and hit Apply,
    Wait few minutes, after the Power LED will stop blinking, the router is ready for configuration.

Known issues
Some USB 2.0 devices work at full speed mode 1.1 only

Additional information
OEM firmware shell password is: SitecomSenao
useful for creating backup of original firmware.
There is also another revision of this device (v1 001), based on RT3352 SoC

Signed-off-by: Andrea Poletti <polex73@yahoo.it>

mpratt14 commented:
spelling of "ramips"

mpratt14 commented:
you have dashes in the "compatible" so they should be in the DTS file name and makefile

mpratt14 commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -964,6 +964,20 @@ define Device/sercomm_na930
+  DEVICE_VENDOR := Sitecom
+  DEVICE_MODEL := WLR-4100
+  DEVICE_VARIANT := v1002
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci
```
I believe you have to add `uboot-envtools` to packages list in order to have it by default

mpratt14 commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -17,7 +17,8 @@ alfa-network,awusfree1|\
 alfa-network,r36m-e4g|\
 alfa-network,tube-e4g|\
-engenius,esr600h)
+engenius,esr600|\
+sitecom,wlr-4100-v1-002)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x1000"
 	;;
```
note the two different models here `esr600` and `esr600h`

polex reviewed (commented):
> you have dashes in the "compatible" so they should be in the DTS file name and makefile
By renaming the file name to mt7620a_sitecom_wlr-4100-v1-002.dts, the comments reference to the original file have been lost.

adschm commented on target/linux/ramips/dts/mt7620a_sitecom_wlr-4100-v1-002.dts:
```diff
@@ -0,0 +1,184 @@
+		wifi {
+			label = "blue:wifi";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
```
How can this be phy1tpt if there only is one radio?

adschm commented:
Since you have both binary and ascii MAC address sources, would you please compile a MAC address table like here (include ascii sources from uboot):

https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

Just post the result here, I've already picked the commit and don't want to redo that again. Current state:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/wlr4100

polex commented:
> Since you have both binary and ascii MAC address sources, would you please compile a MAC address table like here (include ascii sources from uboot):
> 
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses
> 
> Just post the result here, I've already picked the commit and don't want to redo that again. Current state:
> https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/wlr4100

I identified 4 MAC addresses using `printenv` boot command (MTD1 "config"):
1. ethaddr
2. wanaddr
3. wlanaddr
4. iNICaddr

1 and 2 differ in the last digit (...:84 and ...:85). Instead 3 and 4 look dummy.

```
MT7620 # printenv
bootcmd=tftp
bootdelay=1
baudrate=115200
ethaddr="xx:xx:xx:xx:B9:84"
sn=xxxxxxxxx
hw_ver=1.0.0
hw_id=0222104A
pro_id=000
country=000
domain=1
wanaddr=xx:xx:xx:xx:B9:85
wlanaddr=00:AA:BB:CC:DD:12
iNICaddr=00:AA:BB:CC:DD:22
eth_en=0
snextra=xxxxxxxxx***********
uboot_ver=1.0.0.1
sku_table=0
language_code=uk
op_mode=0
filesize=366f82
fileaddr=80A00000
ipaddr=192.168.99.199
serverip=192.168.99.8
autostart=no
bootfile=uImageWLR-5100v1002.bin
stdin=serial
stdout=serial
stderr=serial

Environment size: 482/4092 bytes
```
Instead I identified only ethaddr in factory (MTD2) partition:
```
># hexdump -C mtd2.bin 
00000000  20 76 04 01 64 d1 a3 0f  b9 84 ff ff ff ff ff ff  | v..d...........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  ff ff ff ff ff ff ff ff  00 0c 43 76 20 77 00 0c  |..........Cv w..|

```
The `ifconfig -a` on factory firmware shows just ethaddr address.

mpratt14 commented:
in this case you should use only DTS to take the MACs from factory partition, the uboot environment workaround is only for when no MAC can be found in the other flash partitions, or if the device uses 2 very different MAC addresses, and one of them is not found anywhere else

adschm commented:
Okay, I've merged the new information and will merge this later this week.

polex commented:
> the uboot environment workaround is only for when no MAC can be found in the other flash partitions
The uboot environment as workaround makes it useless in WLR-4100 since we already have the MAC in Firmware. 
I believe it was an added value. Sorry.
